cedar-lean/Cedar/Thm/Partial/Evaluation.lean

```diff
@@ -444,3 +444,114 @@ theorem subst_preserves_evaluation_to_value {expr : Partial.Expr} {req req' : Pa
+    }
+  case or x₁ x₂ =>
+    apply (AndOr.subst_preserves_errors (by intro x _ v ; exact subst_preserves_evaluation_to_value wf_r wf_e h_req) _ _).right h_req e h₁
+    all_goals {
+      unfold SubstPreservesEvaluationToError
+      intro _ e'
+      exact subst_preserves_errors wf_r wf_e h_req
+    }
+  case ite x₁ x₂ x₃ =>
+    apply Ite.subst_preserves_errors (by intro x _ v ; exact subst_preserves_evaluation_to_value wf_r wf_e h_req) _ _ _ h_req e h₁
+    all_goals {
+      unfold SubstPreservesEvaluationToError
+      intro _ e'
+      exact subst_preserves_errors wf_r wf_e h_req
+    }
+  case binaryApp op x₁ x₂ =>
+    apply Binary.subst_preserves_errors (by intro x _ v ; exact subst_preserves_evaluation_to_value wf_r wf_e h_req) _ _ h_req e h₁
+    all_goals {
+      unfold SubstPreservesEvaluationToError
+      intro _ e'
+      exact subst_preserves_errors wf_r wf_e h_req
+    }
+  case unaryApp op x₁ =>
+    apply Unary.subst_preserves_errors (by intro x _ v ; exact subst_preserves_evaluation_to_value wf_r wf_e h_req) _ h_req e h₁
+    all_goals {
+      unfold SubstPreservesEvaluationToError
+      intro _ e'
+      exact subst_preserves_errors wf_r wf_e h_req
+    }
+  case getAttr x₁ attr =>
+    apply GetAttr.subst_preserves_errors wf_e (by intro x _ v ; exact subst_preserves_evaluation_to_value wf_r wf_e h_req) _ h_req e h₁
+    all_goals {
+      unfold SubstPreservesEvaluationToError
+      intro _ e'
+      exact subst_preserves_errors wf_r wf_e h_req
+    }
+  case hasAttr x₁ attr =>
+    apply HasAttr.subst_preserves_errors (by intro x _ v ; exact subst_preserves_evaluation_to_value wf_r wf_e h_req) _ h_req e h₁
+    all_goals {
+      unfold SubstPreservesEvaluationToError
+      intro _ e'
+      exact subst_preserves_errors wf_r wf_e h_req
+    }
+  case set xs =>
+    apply Set.subst_preserves_errors _ h_req e h₁
+    all_goals {
+      unfold SubstPreservesEvaluationToError
+      intro x hx _ e'
+      have := List.sizeOf_lt_of_mem hx
+      exact subst_preserves_errors wf_r wf_e h_req
+    }
+  case record attrs =>
+    apply Record.subst_preserves_errors _ h_req e h₁
+    all_goals {
+      unfold SubstPreservesEvaluationToError
+      intro x hx _ e'
+      have := Map.sizeOf_lt_of_value hx
+      exact subst_preserves_errors wf_r wf_e h_req
+    }
+  case call xfn xs =>
+    apply Call.subst_preserves_errors (by intro x _ v ; exact subst_preserves_evaluation_to_value wf_r wf_e h_req) _ h_req e h₁
+    all_goals {
+      unfold SubstPreservesEvaluationToError
+      intro x hx _ e'
+      have := List.sizeOf_lt_of_mem hx
+      exact subst_preserves_errors wf_r wf_e h_req
+    }
+termination_by expr
+
+/--
+  Corollary (contrapositive) of the above:
+  If partial evaluation returns ok after any substitution of unknowns,
+  then it must return ok before that substitution
+-/
+theorem subst_preserves_errors_mt {expr : Partial.Expr} {req req' : Partial.Request} {entities : Partial.Entities} {subsmap : Subsmap}
```

Comment:
Why `_mt`?

Reply:
From the Latin "modus tollens", which in formal logic is the rule for reducing contrapositives.  I named this in analogue to Lean's use of `mp` (from the Latin "modus ponens") to mean the forward direction of an implication.  Suggestions for a different name? I suppose `_contrapositive`?